pennylane/circuit_graph.py

```diff
@@ -104,14 +104,18 @@ class CircuitGraph:
             The dictionary key is the parameter index.
     """
 
+    # pylint: disable=too-many-public-methods
+
     def __init__(self, ops, variable_deps):
         self.variable_deps = variable_deps
 
         self._grid = {}
         """dict[int, list[Operator]]: dictionary representing the quantum circuit as a grid.
         Here, the key is the wire number, and the value is a list containing the operators on that wire.
         """
+        self.num_wires = 0
         for k, op in enumerate(ops):
+            self.num_wires = max(self.num_wires, max(list(_flatten(op.wires))) + 1)
```

Comment:
Yes! I was shocked as well. I thought this would be a minor addition, and I caused half of the test suite to fail 😆 

Turns out the issue is that tensors have wires which are nested lists:

```python
>>> Tensor(qml.PauliX(0), qml.Hermitian(A, [1, 2])).wires
[[0], [1, 2]]
```

This is definitely something that a `Wires` class will help fix.

Reply:
I think a wires class will help significantly simplify the `Tensor` class

Reply:
That's a great idea! So this represents the number of wires the Operation/Observable **actually** acts on right?
```python
import pennylane as qml
import numpy as np

print(qml.Hermitian.num_wires)

PauliZ_with_hermitian = qml.Hermitian(np.array([[1,0],[0,-1]]), wires=[0])
print(PauliZ_with_hermitian.wires)
```
outputs
```python
ActsOn.AnyWires
[0]
```
So we are looking for `PauliZ_with_hermitian.num_wires=1` in this case.

However, this is unique to the instance and we wouldn't be able to tell with `num_wires` attribute being a class attribute. Seems like a similar case of instance/class attributes with `matrix` and `eigvals`... :thinking:

Reply:
Nitpick: there are two things done now by the same `for` loop:
* choosing the maximum number of wires
* setting the default for the grid

Although there is an advantage that looping is done only once, separating these into two `for` loops could help separate these logical tasks as well.